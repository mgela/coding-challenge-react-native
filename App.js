import {
  StackNavigator,
  addNavigationHelpers,
} from 'react-navigation';
import {
  createStore,
  applyMiddleware,
  combineReducers,
} from 'redux';
import { createReduxBoundAddListener } from 'react-navigation-redux-helpers';
import { Provider, connect } from 'react-redux';
import  store  from './store'
import {AppNavigator}   from './reducers/index'

import React from 'react';
import { StyleSheet, Text, View } from 'react-native';






const addListener = createReduxBoundAddListener("root");

const AppWithNavigationState = connect(state => ({
    nav: state.nav,
}))(({ dispatch, nav }) => (
    <AppNavigator navigation={addNavigationHelpers({ dispatch, state: nav, addListener })} />
));

export default class App extends React.Component {
  render() {
    return (
        <Provider store={store}>
           <AppWithNavigationState/>
        </Provider>

    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
