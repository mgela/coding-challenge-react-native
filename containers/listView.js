/* @flow */

import React, { Component } from 'react';
import {
  View,
  Text,
  StyleSheet,
  FlatList,
} from 'react-native';

import {connect} from 'react-redux';

import {nav, back, articles, paginate} from '../actions'


export class ListView extends Component {
  constructor(props){
    super(props)
    //blocker, onEndReached getting called on render spamming mainview-> listview.
    this.state = {
      flatListReady:false
    }
  }

  //fetch items on component init
  componentDidMount(){
    this.fetchItems()
  }

  fetchItems = async ()=> {
    let totalPages = 0;
    let totalArticles = 0;
    const data = await fetch('https://demo.wp-api.org/wp-json/wp/v2/posts?per_page=10', {
      method: 'GET'
    })
    //get information from headers
    totalPages = data.headers.map["x-wp-totalpages"]["0"]
    totalArticles = data.headers.map["x-wp-total"]["0"]
    const json = await data.json()
    await console.log(json, 'jsondata');
    //remove non useful info
    const structureData = json.map(item => (
      { id: item.id, title: item.title.rendered, content: item.excerpt.rendered.replace(/(<([^>]+)>)/ig, '')}
    ))
    //avoid recalls when first page already fetched
    if (this.props.articlesList.length === 0){
        await this.props.articles(structureData, totalPages, totalArticles)
    }
  }

  //sorry flatlist acting funky
  onScroll = ()=> {
    this.setState({flatListReady:true})
  }

  paginate = async () => {
    try {
      // blockers for preventing onEndReached triggered
        if (!this.state.flatListReady) return null
        else {
          const fetchData = await fetch(`https://demo.wp-api.org/wp-json/wp/v2/posts?page=${this.props.nextPagePagination }&per_page=10`, {method: 'GET'})
          const data = await fetchData.json()
          const structureData =  data.map(article => (
            { id: article.id, title: article.title.rendered, content: article.excerpt.rendered.replace(/(<([^>]+)>)/ig, '')}
          ))
          //more blockers preventing onEndReached getting triggered
          if (this.props.totalPages !== this.props.articlesList.length && this.state.flatListReady){
             await this.props.paginate(structureData)
          }
        }
    } catch (e){
        console.log(e);
      }
  }
  render() {
    return (
      <View style={styles.container}>
        {/* could refactor to pure component */}
          <FlatList
            data={this.props.articlesList}
            keyExtractor={(item)=> item.id}
            renderItem={( { item } )=> (
              <View style={styles.listView}>
                <Text style={{fontSize: 18, color: '#2196f3'}}>article ID:{item.id}</Text>
                <Text style={{fontSize: 23, color: '#2196f3'}}>{item.title}</Text>
                <Text style={{fontSize: 12, color: 'black'}}>{item.content}</Text>
              </View>
            )}
            onEndReached={()=>{this.paginate()}}
            onEndThreshold={0.01}
            onScroll={this.onScroll}
          />
      </View>
    );
  }
}

const mapStateToProps = (state) => ({
  nav: state.nav,
  articlesList: state.counterIncrease.actualArticles,
  totalPages : state.counterIncrease.totalPages,
  totalArticles: state.counterIncrease.totalArticles,
  nextPagePagination: state.counterIncrease.nextPage,
})

const mapDispatchToProps = (dispatch) => ({
  enterMainView: ()=> dispatch(nav()),
  resetNav: ()=> dispatch(back()),
  articles: (data, totalPages, totalArticles)=> dispatch(articles(data, totalPages, totalArticles)),
  paginate: (data)=> dispatch(paginate(data)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ListView);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    justifyContent: 'center',
  },
  listView: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: 'auto',
    borderBottomWidth: 2,
    borderTopWidth: 2,
    marginTop: '6%'
  },
});
