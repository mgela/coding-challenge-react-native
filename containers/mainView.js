/* @flow */

import React, { Component } from 'react';
import {
  View,
  Text,
  StyleSheet,
  Button,
  TouchableOpacity,
} from 'react-native';

import {nav} from '../actions'
import {connect} from 'react-redux';

export class MainView extends Component {

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.counterWrapper}>
          <Text style={styles.text}>Number of articles: </Text>
          <Text style={styles.counter}>{this.props.listItems.length}</Text>
          <TouchableOpacity onPress={this.props.enterMainView}>
            <View style={styles.button}>
              <Text style={styles.text}>Enter List</Text>
            </View>
          </TouchableOpacity>
        </View>
        </View>
    );
  }
}

const mapStateToProps = (state) => ({
  nav: state.nav,
  listItems: state.counterIncrease.actualArticles
})
const mapDispatchToProps = (dispatch) => ({
  enterMainView: ()=> dispatch(nav())
})

export default connect(mapStateToProps, mapDispatchToProps)(MainView);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  counterWrapper: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: '100%',
    height: '50%',
  },
  button: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: 200,
    backgroundColor: 'lime',
    borderRadius: 10,
  },
  counter: {
    fontSize: 60,
  },
  text: {
    fontSize: 24,
    color: '#2196f3',
  },

});
