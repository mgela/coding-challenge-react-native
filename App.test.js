import React from 'react';
import App from './App';

import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow'; 


//screens
import {MainView} from './containers/mainView'
import {ListView} from './containers/listView'

it('renders without crashing', () => {
  const rendered = renderer.create(<App />).toJSON();
  expect(rendered).toBeTruthy();
});


test('renders correctly', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<ListView />);
const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
test('renders correctly', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<ListView />);
const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
