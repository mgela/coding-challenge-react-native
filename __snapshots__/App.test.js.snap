// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <FlatList
    data={undefined}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onEndThreshold={0.01}
    onScroll={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`renders correctly 2`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <FlatList
    data={undefined}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onEndThreshold={0.01}
    onScroll={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
